test/unit-test/FreeRTOS_DHCP/FreeRTOS_DHCP_utest.c

```diff
@@ -0,0 +1,3657 @@
+    TEST_ASSERT_EQUAL( &xTestSocket, xDHCPSocket );
+    /* The state should indicate that we still in the state from where we started. */
+    TEST_ASSERT_EQUAL( eWaitingOffer, xDHCPData.eDHCPState );
+    /* make sure that the period is increased by a factor of two. */
+    TEST_ASSERT_EQUAL( ( ( ipconfigMAXIMUM_DISCOVER_TX_PERIOD >> 1 ) - 1 ) << 1, xDHCPData.xDHCPTxPeriod );
+}
+
+void test_vDHCPProcess_eWaitingOfferRecvfromFailsTimeoutDontGiveUpRNGPassNoBroadcast( void )
+{
+    struct xSOCKET xTestSocket;
+    TickType_t xTimeValue = 1234;
+
+    /* This should remain unchanged. */
+    xDHCPSocket = &xTestSocket;
+    /* Put the required state. */
+    xDHCPData.eDHCPState = eWaitingOffer;
+    /* Put some time values. */
+    xDHCPData.xDHCPTxTime = 100;
+    /* Make sure that we don't exceed the period - and thus, don't give up. */
+    xDHCPData.xDHCPTxPeriod = ( ipconfigMAXIMUM_DISCOVER_TX_PERIOD >> 1 ) - 1;
+    /* Not Using broadcast. */
+    xDHCPData.xUseBroadcast = pdFALSE;
+
+    /* Expect these arguments. Return a 0 to fail. */
+    FreeRTOS_recvfrom_ExpectAndReturn( xDHCPSocket, NULL, 0UL, FREERTOS_ZERO_COPY, NULL, NULL, 0 );
+    /* Ignore the buffer argument though. */
+    FreeRTOS_recvfrom_IgnoreArg_pvBuffer();
+
+    /* Make sure that there is timeout. The expression is: xTaskGetTickCount() - EP_DHCPData.xDHCPTxTime ) > EP_DHCPData.xDHCPTxPeriod  */
+    /* Return a value which makes the difference greater than the period. */
+    xTaskGetTickCount_ExpectAndReturn( xDHCPData.xDHCPTxTime + xDHCPData.xDHCPTxPeriod + 100 );
+
+    /* Make all calls to the RNG succeed. */
+    xApplicationGetRandomNumber_IgnoreAndReturn( pdTRUE );
+
+    xTaskGetTickCount_ExpectAndReturn( xTimeValue );
+
+    /* Get the hostname. */
+    pcApplicationHostnameHook_ExpectAndReturn( pcHostName );
+    /* Returning a NULL network buffer. */
+    pxGetNetworkBufferWithDescriptor_IgnoreAndReturn( NULL );
+
+    vDHCPProcess( pdFALSE, eWaitingOffer );
+
+    /* DHCP socket should be unallocated */
+    TEST_ASSERT_EQUAL( &xTestSocket, xDHCPSocket );
+    /* The state should indicate that we still in the state from where we started. */
+    TEST_ASSERT_EQUAL( eInitialWait, xDHCPData.eDHCPState );
+    /* make sure that the period is increased by a factor of two. */
+    TEST_ASSERT_EQUAL( ( ( ipconfigMAXIMUM_DISCOVER_TX_PERIOD >> 1 ) - 1 ) << 1, xDHCPData.xDHCPTxPeriod );
+}
+
+static int32_t FreeRTOS_recvfrom_eWaitingOfferRecvfromSucceedsFalseCookieNoTimeout( Socket_t xSocket,
+                                                                                    void * pvBuffer,
+                                                                                    size_t uxBufferLength,
+                                                                                    BaseType_t xFlags,
+                                                                                    struct freertos_sockaddr * pxSourceAddress,
```

Comment:
cleaner if these are at the top... if the file is too large, maybe better to split it

Reply:
I concur. Done.